Added interfaces to wifi

zanaviska opened the pull request:
# issue
Currently, if there are several wireless interfaces, there is no way to figure it out, which connection is being used(see picture below. There is no way to determine difference between 1 and 2 or 3 and 4 or 5 and 6).  
In addition to that, when trying to connect to some ssid, it will ALWAYS use the first wireless interface and so, there is no way to use the second one. So there is no actual way to connect to 2, only to 1
![image](https://github.com/user-attachments/assets/e6c3b153-87bf-416f-932e-c4633629fe62)

# fix
Fix contains several changes:
1) when connecting to wifi, it compares both ssid and iface of the connected wifi
2) it displays iface for ssid that is available on several interfaces
3) same rule applies to already connected wifi(see last image)
![image](https://github.com/user-attachments/assets/7cab976a-478f-48e5-af6f-07bd01c5de24)
![image](https://github.com/user-attachments/assets/18199687-a007-464a-a85b-2ee975adf659)

mmstick commented:
This'll require design input

zanaviska commented:
Has there been any progress on the design input since the last message?

maria-komarova commented:
Sorry for the delay, but we have just discussed adjusting designs for multiple adapters for wireless and wired connections last week. The applet designs are described in #623 
Here is the screenshot of the wireless adapters. I used the same IDs you had so that it is easier to see the suggestions: 
![image](https://github.com/user-attachments/assets/bac200ad-688c-41ef-90a1-0788587ed926)
The idea is to group the connections by the adapter and then display them inside the nested popup instead of in one giant list. 
I am not sure if we already have the nested popup example anywhere, that might still be pending implementation.

Dygear commented:
> Sorry for the delay, but we have just discussed adjusting designs for multiple adapters for wireless and wired connections last week. The applet designs are described in #623 Here is the screenshot of the wireless adapters. I used the same IDs you had so that it is easier to see the suggestions: [snip] The idea is to group the connections by the adapter and then display them inside the side popup menu instead of in one giant list.

That looks good. My critique is that the top level should probably have the IP address instead of just the connected label. Might as well show that information up front. No need to go digging into the menu (even if it is just one layer deeper.) Looks like that detail was included in the #623 figma for the single adapter instance.

Maybe a friendly device name option for the interface, so instead of `wlo1`, `wlx503eaa3bbf11`, and `adaptor3` if it's easily possible to pluck the manufacture's name then the last 3 octets of the MAC address. For Intel it would be `Intel DE:AD:BE`. And `RealTek EF:13:37`. As the first three octets of the MAC address are manufacture prefixes anyway, so omitting them won't matter much and it allows you to localize down to what adapter you want to use much more easily.

zanaviska commented:
@maria-komarova I have some question about how should it work from User perspective(all names is based on your example):
1) In case when **wlo1** and **wlx503...** have known networks, but **adapter3** has no known network, we will still display adapters together as on your screen?
2) If I am connected to **adapter3**, should I write ip address of a connection?
3) If we are viewing wifi networks related to some adapter, how should we mark "connected" wifi? "known" wifi? Should we draw them in separate block? Should we write ip address of a connected network? Or how should it look?

zanaviska commented:
Hello. @maria-komarova I have made such an implementation. It is still
buggy and is yet to be ready for the review, but I think you could already
tell me if I am doing smt wrong from design perspective


пт, 23 серп. 2024 р. о 18:49 maria-komarova ***@***.***> пише:

> Sorry for the delay, but we have just discussed adjusting designs for
> multiple adapters for wireless and wired connections last week. The applet
> designs are described in #623
> <https://github.com/pop-os/cosmic-applets/issues/623>
> Here is the screenshot of the wireless adapters. I used the same IDs you
> had so that it is easier to see the suggestions:
> image.png (view on web)
> <https://github.com/user-attachments/assets/bac200ad-688c-41ef-90a1-0788587ed926>
> The idea is to group the connections by the adapter and then display them
> inside the nested popup instead of in one giant list.
> I am not sure if we already have the nested popup example anywhere, that
> might still be pending implementation.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pop-os/cosmic-applets/pull/584#issuecomment-2307356531>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSGL53QTDD6ROY5V2GRKSLZS5KXZAVCNFSM6AAAAABMOCV5OSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBXGM2TMNJTGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

zanaviska commented:
чт, 24 жовт. 2024 р. о 15:46 Богдан Зіновійович ***@***.***> пише:

> Hello. @maria-komarova I have made such an implementation. It is still
> buggy and is yet to be ready for the review, but I think you could already
> tell me if I am doing smt wrong from design perspective
>
>
> пт, 23 серп. 2024 р. о 18:49 maria-komarova ***@***.***>
> пише:
>
>> Sorry for the delay, but we have just discussed adjusting designs for
>> multiple adapters for wireless and wired connections last week. The applet
>> designs are described in #623
>> <https://github.com/pop-os/cosmic-applets/issues/623>
>> Here is the screenshot of the wireless adapters. I used the same IDs you
>> had so that it is easier to see the suggestions:
>> image.png (view on web)
>> <https://github.com/user-attachments/assets/bac200ad-688c-41ef-90a1-0788587ed926>
>> The idea is to group the connections by the adapter and then display them
>> inside the nested popup instead of in one giant list.
>> I am not sure if we already have the nested popup example anywhere, that
>> might still be pending implementation.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/pop-os/cosmic-applets/pull/584#issuecomment-2307356531>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AHSGL53QTDD6ROY5V2GRKSLZS5KXZAVCNFSM6AAAAABMOCV5OSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBXGM2TMNJTGE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

zanaviska commented:
[image: wifi_beta(1).gif]

чт, 24 жовт. 2024 р. о 15:47 Богдан Зіновійович ***@***.***> пише:

>
>
> чт, 24 жовт. 2024 р. о 15:46 Богдан Зіновійович ***@***.***>
> пише:
>
>> Hello. @maria-komarova I have made such an implementation. It is still
>> buggy and is yet to be ready for the review, but I think you could already
>> tell me if I am doing smt wrong from design perspective
>>
>>
>> пт, 23 серп. 2024 р. о 18:49 maria-komarova ***@***.***>
>> пише:
>>
>>> Sorry for the delay, but we have just discussed adjusting designs for
>>> multiple adapters for wireless and wired connections last week. The applet
>>> designs are described in #623
>>> <https://github.com/pop-os/cosmic-applets/issues/623>
>>> Here is the screenshot of the wireless adapters. I used the same IDs you
>>> had so that it is easier to see the suggestions:
>>> image.png (view on web)
>>> <https://github.com/user-attachments/assets/bac200ad-688c-41ef-90a1-0788587ed926>
>>> The idea is to group the connections by the adapter and then display
>>> them inside the nested popup instead of in one giant list.
>>> I am not sure if we already have the nested popup example anywhere, that
>>> might still be pending implementation.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/pop-os/cosmic-applets/pull/584#issuecomment-2307356531>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AHSGL53QTDD6ROY5V2GRKSLZS5KXZAVCNFSM6AAAAABMOCV5OSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMBXGM2TMNJTGE>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>

maria-komarova commented:
Looks good from the design perspective so far, just buggy.

zanaviska commented:
https://github.com/user-attachments/assets/ba7dd6ba-daed-4e8d-b0d3-194d3d273939

The UI part is ready. However on lower level there are 3 bugs left: one causes a crash that can be observed at the end of a video and other occur when interaction with one device affects other one

Regarding displaying non-raw MAC addresses but rather manufacturer names. It could be done, but I need confirmation where should I get the list of manufacturer and mac-prefixes from. I found some list on [cisco website](https://devtools360.com/en/macaddress/vendorMacs.xml), on some random [github gist](https://gist.github.com/aallan/b4bb86db86079509e6159810ae9bd3e4) or this [gist](https://gist.github.com/NullArray/0380871a42b608830357f998df735e71). However I am unsure if any of these sources have licenses or/and could be use as a part of Cosmic DE.
Another thing to mention about such proposal is naming. All mentioned lists could offer lengthy names for vendors. For example in my case name would look like this. 
![image](https://github.com/user-attachments/assets/cfabade5-cd3f-461d-92b1-dc30c2b0ba56)
![image](https://github.com/user-attachments/assets/446841bd-4352-492d-b4ed-164eaeaf56e8)

zanaviska commented:
I'd like to bring this PR back for a review. Overall it is ready. There are 3 TODOs left, but I don't thing I know your codebase good enough to handle them as well as they have little to no impact on the user. Some technical info about them:
1) in **app.rs** `// TODO: remove excesive column!`
In this place I have added excessive `column!` because toggle(in our case wifi toggle) directly after separator could create a `panic!` which terminates the program. So this is workaround over some bug in toggle-widget(most likely). I have spent several days researching the bug, but this is best I could find out. If you want more details I could give them to you(or even give steps how to reproduce it on your machine)
2) in **mod.rs** I have added 2 todos
I do not know how to get MAC address for `cosmic_dbus_networkmanager::settings::connection::Connection`. It could affect user only in cases when user either try to connect to wifi with ssid same as ssid that other wifi card is connected to or if forgetting connection on one wifi device could forget it for another also(yet this is consistent with "learning" WiFis)

zanaviska commented:
@mmstick Hello. I am done on my side. It is ready for a review. I do not know how your workflow work, but I would like to know about any status update regarding this PR. 

And the question about MAC addresses asked [here](https://github.com/pop-os/cosmic-applets/pull/584#issuecomment-2524093183) still exist

wash2 commented:
@zanaviska Unfortunately, I don't believe there is a great way of getting the mac address from the dbus interface. You can get the mac address for a given device at `/sys/class/net/{device interface}/address`. Otherwise, this looks good to me after it passes the formatting check.

zanaviska commented:
It now passes formatting check

wash2 commented:
> It now passes formatting check @wash2 I didn't get your answer about MAC addresses. Current version of PR relies on MAC address. Since there was a [suggestion](https://github.com/pop-os/cosmic-applets/pull/584#issuecomment-2330748495) to use MAC addresses. I no longer use interface names. The question, which still is in pending for an answer is about a way of getting manufacturer name for given MAC-address
> 
> > Regarding displaying non-raw MAC addresses but rather manufacturer names. It could be done, but I need confirmation where should I get the list of manufacturer and mac-prefixes from. I found some list on [cisco website](https://devtools360.com/en/macaddress/vendorMacs.xml), on some random [github gist](https://gist.github.com/aallan/b4bb86db86079509e6159810ae9bd3e4) or this [gist](https://gist.github.com/NullArray/0380871a42b608830357f998df735e71). However I am unsure if any of these sources have licenses or/and could be use as a part of Cosmic DE.
> > Another thing to mention about such proposal is naming. All mentioned lists could offer lengthy names for vendors. For example in my case name would look like this.

This is probably ok for now. I think you might need to look at `/sys/class/net/{interface name}/device/vendor` and `/sys/class/net/{interface name}/device/device` for the manufacturer ids. And check `/usr/share/misc/pci.ids` for the actual name given the prefix, but this path would probably need to be configurable.
